BUG: Fix numerical precision error of `truncnorm.logcdf` when cdf is nearly 1

not522 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
<!--Please explain your changes.-->

Resolve the large precision error of `truncnorm.logcdf` with upper tail argument. `mass_case_central` of `_log_gauss_mass` was inaccurate due to catastrophic cancellation in `logsumexp`.

#### Additional information
<!--Any additional information you think is important.-->

- Relative error using SciPy v1.9.1

![truncnorm_logcdf_rel_error_1 9 1](https://user-images.githubusercontent.com/1255726/191490249-c1259b1c-21de-4e85-be77-b0ff986febc7.png)

- Relative error using this PR

![truncnorm_logcdf_rel_error_pr](https://user-images.githubusercontent.com/1255726/191490284-2c9c85fc-ed96-45b4-8cd9-de779173c066.png)

- Reproducible code (based on #15172)

```python
import numpy as np
import scipy
from scipy.stats import truncnorm
import mpmath
import matplotlib.pyplot as plt


def truncnorm_logcdf_mp(x, a, b):
    ndtr_x = mpmath.ncdf(x)
    ndtr_a = mpmath.ncdf(a)
    ndtr_b = mpmath.ncdf(b)
    return mpmath.log((ndtr_x - ndtr_a) / (ndtr_b - ndtr_a))


mpmath.mp.dps = 100

clips = [(-float("inf"), float("inf")), (-10, 10), (-5, 5), (-1, 1)]

fig, axes = plt.subplots(2, len(clips)//2, figsize=(9, 5), constrained_layout=True)

for (a, b), axis in zip(clips, axes.ravel()):
    if (a, b) == (-float("inf"), float("inf")):
        x = np.linspace(-10, 10, 1250, endpoint=False)
    else:
        x = np.linspace(a, b, 1250, endpoint=False)
    ymp = np.array([truncnorm_logcdf_mp(t, a, b) for t in x])
    y = truncnorm.logcdf(x, a, b)
    relerr = np.abs((y - ymp) / ymp)

    axis.plot(x, relerr)
    axis.set_yscale('log')
    axis.grid()
    axis.set_title(f'logcdf(x, {a}, {b})')

fig.suptitle('Relative error of truncnorm.logcdf(x), '
             f'SciPy version {scipy.__version__}\n ')
plt.savefig(f'truncnorm_logcdf_rel_error_{scipy.__version__}.png')
```

mdhaber commented:
When we worked on this recently, we focused on the case of small CDF. I suppose I missed this when the CDF is nearly 1. I'll take a look shortly.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8590,7 +8590,16 @@ def mass_case_right(a, b):
     def mass_case_central(a, b):
         left_mass = mass_case_left(a, 0.5)
         right_mass = mass_case_right(0.5, b)
```
```suggestion
        left_mass = mass_case_left(a, 0)
        right_mass = mass_case_right(0, b)
```

mdhaber commented:
Thanks @not522. I performed some experiments similar to those described in https://github.com/scipy/scipy/pull/16605#issuecomment-1186277384. In that space of possible inputs, it actually didn't run into cases for which this alternative formulation substantially improves the accuracy, but it did find some cases in which this new formulation is a bit less accurate than the original. 

To get the best of both worlds, I adjusted this to use this new formulation only when needed - when the log of the probability mass is nearly 0. I also corrected some mistakes I saw in the original code - those `0.5` were because I was thinking of the CDF being 0.5, but that corresponds with the argument of the function being 0. I'd been meaning to fix that.

If this looks good to you, I'll go ahead and merge.

not522 commented:
Thank you for your reply.

> I performed some experiments similar to those described in #16605 (comment) but tailored toward the case in which a < 0 and b > 0. In that space of possible inputs, it actually didn't run into cases for which this alternative formulation substantially improves the accuracy, but it did find some cases in which this new formulation is a bit less accurate than the original. (Of course, in the cases that this formulation is needed, it really improves the accuracy.)

I think this error does not affect `logpdf`. I confirmed `logcdf` has a large relative error using your script.

> To get the best of both worlds, I adjusted this to use this new formulation only when needed - when the log of the probability mass is nearly 0.

If the exact way depends on the input, I agree with your opinion, but I have not been able to identify any cases where the previous methods are suitable. Could you give me explicit input?

> I also corrected some mistakes I saw in the original code - those 0.5 were because I was thinking of the CDF being 0.5, but that corresponds with the argument of the function being 0. I'd been meaning to fix that.

nitpicky: You missed 0.5 in `_log_diff`.

steppi commented:
> @steppi perhaps this is the catastrophic cancellation you were looking for?

Sorry, when I reviewed gh-16605 I'd assumed that `log_diff` was essentially using the trick from my comment [here](https://github.com/scipy/scipy/issues/13923#issuecomment-825118292). Checking the implementation of [logsumexp](https://github.com/scipy/scipy/blob/606e84b0e4370c885eec4b7acca5ec00797b7d98/scipy/special/_logsumexp.py#L7), it seems to be doing something else. I won't have time to look more carefully until the first week of October though. I used the `log_diff` from my comment to implement the log cdf of the truncated Beta distribution and it worked well enough for my purposes. Even there I saw some catastrophic cancellation though.

I'll have a chance to look closer in a few weeks.

mdhaber commented:
There was another source of catastrophic cancellation.

When we developed this (main), we were focused on substantially negative values of `logcdf` (because I was really focused on `cdf`, and bad relative errors in near-zero values of `logcdf` don't show up in `cdf`). The original fix patched the behavior for nearly-zero values of `logcdf` when the distribution is lightly truncated. The more recent patches fix the behavior for nearly-zero values of `logcdf` (and `logsf`) when the distribution is heavily truncated.

main:
![image](https://user-images.githubusercontent.com/6570539/191776577-5d20c8ae-4c91-4478-a6b3-7dcf72a65f4d.png)

previous commit:
![image](https://user-images.githubusercontent.com/6570539/191776966-06263a0e-7281-4ac0-95e5-2502aad7b7e1.png)

Now:
![image](https://user-images.githubusercontent.com/6570539/191776133-5d66266e-9dd7-4857-81a1-0f1100c4775f.png)

We can still lose precision for nearly-zero values of `logcdf` and `logsf` when one of the truncation points is near 0 (e.g. halfnormal) because `logndtr` for the numerator and denominator are nearly the same (and evaluating in the other tail doesn't help). I have some ideas for that, but I'd like to do it in another PR so I don't keep pushing to @not522's branch.

@not522, if it sounds good to you, I'll rebase to squash all of _my_ commits into one, force-push to your branch, then merge your commit and my one commit. I'll keep a backup of the state before the rebase, but you might want to do the same if you care about preserving history. Does that sound OK to you?

While it's still not perfect, I do think this is a big improvement for `logcdf` near zero, and the code is still relatively simple.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8565,20 +8565,19 @@ def _log_sum(log_p, log_q):
 # same as above, but using -exp(x) = exp(x + πi)
 def _log_diff(log_p, log_q):
-    # need to broadcast in case a or b is 0.5; logsumexp doesn't
+    # need to broadcast in case a or b is 0; logsumexp doesn't
     log_p, log_q = np.broadcast_arrays(log_p, log_q)
```
This is probably not needed anymore due to the change in `mass_case_central`, but I'll remove it later.

mdhaber commented on scipy/stats/tests/test_continuous_basic.py:
```diff
@@ -401,18 +401,18 @@ def test_nomodify_gh9900_regression():
     # Check that the cdf and _cdf return the same result.
     npt.assert_almost_equal(tn.cdf(1, 0, np.inf), 0.6826894921370859)
-    npt.assert_almost_equal(tn._cdf(1, 0, np.inf), 0.6826894921370859)
+    npt.assert_almost_equal(tn._cdf([1], [0], [np.inf]), 0.6826894921370859)
```
All `rv_continuous` private methods should be able to assume they are getting arrays that are at least 1d because that's what they are passed from the public method. The implementation of `_logcdf` relies on this, so rather than add an additional `np.atleast_1d` call in `_logcdf`, ensure the input is at least 1d here.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8648,13 +8655,24 @@ def _cdf(self, x, a, b):
         return np.exp(self._logcdf(x, a, b))
 
     def _logcdf(self, x, a, b):
-        return _log_gauss_mass(a, x) - _log_gauss_mass(a, b)
```
This was another source of catastrophic cancellation - when the terms are approximately equal but not nearly zero. When this difference of the two terms is nearly zero (`>-0.1`, say), we avoid the catastropic cancelation by working through the survival function. We implement the same idea in `_logsf` (but avoid infinite back and forth calls between the two functions by choosing `-0.1 > log(1/2)`).

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8565,20 +8565,19 @@ def _log_sum(log_p, log_q):
     log_p, log_q = np.broadcast_arrays(log_p, log_q)
     return sc.logsumexp([log_p, log_q+np.pi*1j], axis=0)
 
-
```
Oops. If linting complains, I'll add this back.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -8648,13 +8655,24 @@ def _cdf(self, x, a, b):
 
     def _logsf(self, x, a, b):
-        return _log_gauss_mass(x, b) - _log_gauss_mass(a, b)
+        # return _log_gauss_mass(x, b) - _log_gauss_mass(a, b)
```
```suggestion
```

not522 commented:
I checked your changes, and the result is slightly different from the attached image.

![truncnorm_logcdf_rel_error_pr2](https://user-images.githubusercontent.com/1255726/191909270-ce816be5-d9d9-4529-9947-161b01ab7ff2.png)

Anyway, I agree this improves the accuracy of `logcdf`. I do not mind git history and follow scipy's manner. Please go ahead.

mdhaber commented:
Thanks @not522!
